src/ddoc_cache/src/ddoc_cache_lru.erl

```diff
@@ -0,0 +1,316 @@
+    {ok, #st{
+        atimes = ATimes,
+        dbs = Dbs,
+        time = 0,
+        evictor = Evictor
+    }}.
+
+
+terminate(_Reason, St) ->
+    case is_pid(St#st.evictor) of
+        true -> exit(St#st.evictor, kill);
+        false -> ok
+    end,
+    ok.
+
+
+handle_call({insert, Key, Val}, _From, St) ->
+    #st{
+        atimes = ATimes,
+        dbs = Dbs,
+        time = Time
+    } = St,
+    NewTime = Time + 1,
+    NewSt = St#st{time = NewTime},
+    Pid = ddoc_cache_refresher:spawn_link(Key, ?REFRESH_TIMEOUT),
+    true = ets:insert(?CACHE, #entry{key = Key, val = Val, pid = Pid}),
+    true = ets:insert(?LRU, {NewTime, Key}),
+    ok = khash:put(ATimes, Key, NewTime),
+    store_key(Dbs, Key),
+    trim(NewSt),
+    ?EVENT(inserted, {Key, Val}),
+    {reply, ok, NewSt};
+
+handle_call({update, Key, Val}, _From, St) ->
+    #st{
+        atimes = ATimes
+    } = St,
+    case khash:lookup(ATimes, Key) of
+        {value, _} ->
+            ets:update_element(?CACHE, Key, {#entry.val, Val}),
+            ?EVENT(updated, {Key, Val}),
+            {reply, ok, St};
+        not_found ->
+            {reply, evicted, St}
+    end;
+
+handle_call({remove, Key}, _From, St) ->
+    #st{
+        atimes = ATimes,
+        dbs = Dbs
+    } = St,
+    case khash:lookup(ATimes, Key) of
+        {value, ATime} ->
+            [#entry{pid = Pid}] = ets:lookup(?CACHE, Key),
+            ddoc_cache_refresher:stop(Pid),
+            remove_key(St, Key, ATime),
+
+            DbName = ddoc_cache_entry:dbname(Key),
+            DDocId = ddoc_cache_entry:ddocid(Key),
+            {value, DDocIds} = khash:lookup(Dbs, DbName),
+            {value, Keys} = khash:lookup(DDocIds, DDocId),
+            ok = khash:del(Keys, Key),
+            case khash:size(Keys) of
+                0 -> khash:del(DDocIds, DDocId);
+                _ -> ok
+            end,
+            case khash:size(DDocIds) of
+                0 -> khash:del(Dbs, DDocId);
+                _ -> ok
+            end,
+
+            ?EVENT(removed, Key);
+        not_found ->
+            ok
+    end,
+    {reply, ok, St};
+
+handle_call(Msg, _From, St) ->
+    {stop, {invalid_call, Msg}, {invalid_call, Msg}, St}.
+
+
+handle_cast({accessed, Key}, St) ->
+    #st{
+        atimes = ATimes,
+        time = Time
+    } = St,
+    NewTime = Time + 1,
+    case khash:lookup(ATimes, Key) of
+        {value, OldTime} ->
+            [#entry{pid = Pid}] = ets:lookup(?CACHE, Key),
+            true = is_process_alive(Pid),
+            true = ets:delete(?LRU, OldTime),
+            true = ets:insert(?LRU, {NewTime, Key}),
+            ok = khash:put(ATimes, Key, NewTime),
+            ?EVENT(accessed, Key);
+        not_found ->
+            % Likely a client read from the cache while an
+            % eviction message was in our mailbox
+            ok
+    end,
+    {noreply, St};
+
+handle_cast({evict, DbName}, St) ->
+    gen_server:abcast(mem3:nodes(), ?MODULE, {do_evict, DbName}),
```

Comment:
Should this be `[node() | mem3:nodes()]` to ensure the `do_evict` logic is triggered on every node? Unless I'm missing something it looks like you'll trigger do_evict on all the other nodes besides this one.